### PR TITLE
GG-38922 Node should check liveness of existing connection when inbound connection occurs

### PR DESCRIPTION
**Pull request by Erixonich**

### CODE REVIEW

File: modules/core/src/main/java/org/apache/ignite/internal/util/nio/GridTcpNioCommunicationClient.java [hunk 1]
```diff
@@ -148,4 +156,23 @@ public GridNioSession session() {
...
+    public void sendHeartbeatsIfNeeded() {
+        if (!useHeartbeats)
+            return;
+
+        long sinceLastHeartbeat = U.currentTimeMillis() - ses.lastHeartbeat();
+
+        if (sinceLastHeartbeat > HEARTBEAT_FREQUENCY) {
+            HeartbeatMessage msg = new HeartbeatMessage();
+
+            //if (log.isDebugEnabled())
```
Comment (Erixonich) [inline modules/core/src/main/java/org/apache/ignite/internal/util/nio/GridTcpNioCommunicationClient.java:169]: I will delete this

File: modules/core/src/main/java/org/apache/ignite/internal/util/nio/GridTcpNioCommunicationClient.java [hunk 2]
```diff
@@ -148,4 +156,23 @@ public GridNioSession session() {
     @Override public String toString() {
         return S.toString(GridTcpNioCommunicationClient.class, this, super.toString());
     }
+
+    public void sendHeartbeatsIfNeeded() {
```
[the remark this answers is not in the document]
Comment (rpuch) [reply modules/core/src/main/java/org/apache/ignite/internal/util/nio/GridTcpNioCommunicationClient.java:160]: Also, the method name is about sending 'heatbeatS', but it actually sends just one heartbeat.

File: modules/core/src/main/java/org/apache/ignite/internal/util/nio/GridTcpNioCommunicationClient.java [hunk 3]
```diff
@@ -148,4 +156,23 @@ public GridNioSession session() {
...
+
+        long sinceLastHeartbeat = U.currentTimeMillis() - ses.lastHeartbeat();
+
+        if (sinceLastHeartbeat > HEARTBEAT_FREQUENCY) {
+            HeartbeatMessage msg = new HeartbeatMessage();
+
+            //if (log.isDebugEnabled())
+                log.info("Heartbeat message sent [rmtAddr=" + ses.remoteAddress() + "], sinceLastHeartbeat=" + sinceLastHeartbeat
+                        + " lastSent=" + ses.lastSendTime() + " lastReceived=" + ses.lastReceiveTime() + " scheduletTime=" + ses.lastSendScheduleTime());
+
```
Comment (rpuch) [inline modules/core/src/main/java/org/apache/ignite/internal/util/nio/GridTcpNioCommunicationClient.java:172]: This method is only called once on first connection message. Shouldn't it be called periodically during all lifetime of a connection?

File: modules/core/src/main/java/org/apache/ignite/spi/communication/tcp/TcpCommunicationSpi.java [hunk 4]
```diff
@@ -318,6 +318,9 @@ public class TcpCommunicationSpi extends TcpCommunicationConfigInitializer {
     /** Handshake wait message type. */
     public static final short HANDSHAKE_WAIT_MSG_TYPE = -28;
 
+    /** Heartbeat message type. */
+    public static final short HEARTBEAT_MSG_TYPE = -42;
```
[the remark this answers is not in the document]
Comment (Erixonich) [reply modules/core/src/main/java/org/apache/ignite/spi/communication/tcp/TcpCommunicationSpi.java:322]: First of all that number was free,
Secondly, that is a good number, don't you think?

File: modules/core/src/main/java/org/apache/ignite/internal/util/nio/GridTcpNioCommunicationClient.java [hunk 5]
```diff
@@ -29,18 +29,24 @@
...
 import org.apache.ignite.plugin.extensions.communication.Message;
+import org.apache.ignite.spi.communication.tcp.messages.HeartbeatMessage;
 import org.jetbrains.annotations.Nullable;
 
 /**
  * Grid client for NIO server.
  */
 public class GridTcpNioCommunicationClient extends GridAbstractCommunicationClient {
+    /** Time in ms between different heartbeat messages. */
+    public static final long HEARTBEAT_FREQUENCY = 2000L;
```
[the remark this answers is not in the document]
Comment (Erixonich) [reply modules/core/src/main/java/org/apache/ignite/internal/util/nio/GridTcpNioCommunicationClient.java:40]: done

File: modules/core/src/main/java/org/apache/ignite/internal/util/nio/GridTcpNioCommunicationClient.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (Erixonich) [reply modules/core/src/main/java/org/apache/ignite/internal/util/nio/GridTcpNioCommunicationClient.java:160]: renamed to `checkConnection`

File: modules/core/src/main/java/org/apache/ignite/spi/communication/tcp/internal/InboundConnectionHandler.java [hunk 6]
```diff
@@ -35,14 +35,7 @@
...
 import org.apache.ignite.internal.util.future.GridFutureAdapter;
-import org.apache.ignite.internal.util.nio.GridCommunicationClient;
-import org.apache.ignite.internal.util.nio.GridNioMessageTracker;
-import org.apache.ignite.internal.util.nio.GridNioRecoveryDescriptor;
-import org.apache.ignite.internal.util.nio.GridNioServerListenerAdapter;
-import org.apache.ignite.internal.util.nio.GridNioSession;
-import org.apache.ignite.internal.util.nio.GridNioSessionMetaKey;
-import org.apache.ignite.internal.util.nio.GridShmemCommunicationClient;
-import org.apache.ignite.internal.util.nio.GridTcpNioCommunicationClient;
+import org.apache.ignite.internal.util.nio.*;
```
[the remark this answers is not in the document]
Comment (Erixonich) [reply modules/core/src/main/java/org/apache/ignite/spi/communication/tcp/internal/InboundConnectionHandler.java:38]: fixed

File: modules/core/src/main/java/org/apache/ignite/internal/util/nio/GridTcpNioCommunicationClient.java [hunk 7]
```diff
@@ -157,22 +161,24 @@ public GridNioSession session() {
         return S.toString(GridTcpNioCommunicationClient.class, this, super.toString());
     }
 
-    public void sendHeartbeatsIfNeeded() {
-        if (!useHeartbeats)
+    /**
+     * Sends special {@link ConnectionCheckMessage} through the channel to check if connection still alive.
+     */
+    public void checkConnection() {
```
Comment (rpuch) [inline modules/core/src/main/java/org/apache/ignite/internal/util/nio/GridTcpNioCommunicationClient.java:167]: Maybe `checkConnectionIfEnabled()` (or something similar, to help a reader understand that it's actually a conditional send)?

File: modules/core/src/main/java/org/apache/ignite/internal/util/nio/GridTcpNioCommunicationClient.java [hunk 8]
```diff
@@ -157,22 +161,24 @@ public GridNioSession session() {
...
+        if (!enableConnectionCheckMessage)
             return;
 
-        long sinceLastHeartbeat = U.currentTimeMillis() - ses.lastHeartbeat();
+        long sinceLastConnectionCheck = U.currentTimeMillis() - lastConnectionCheck;
 
-        if (sinceLastHeartbeat > HEARTBEAT_FREQUENCY) {
-            HeartbeatMessage msg = new HeartbeatMessage();
+        if (sinceLastConnectionCheck > CONNECTION_CHECK_MIN_INTERVAL_MS) {
```
Comment (rpuch) [inline modules/core/src/main/java/org/apache/ignite/internal/util/nio/GridTcpNioCommunicationClient.java:173]: Do we need this throttling at all? This is triggered by an event that is pretty rare (an incoming connection attempt), so it doesn't seem to make much sense to throttle it, it seems.

File: modules/core/src/main/java/org/apache/ignite/internal/util/nio/GridTcpNioCommunicationClient.java (hunk 7 above)
Comment (Erixonich) [reply modules/core/src/main/java/org/apache/ignite/internal/util/nio/GridTcpNioCommunicationClient.java:167]: ok

File: modules/core/src/main/java/org/apache/ignite/internal/util/nio/GridTcpNioCommunicationClient.java (hunk 8 above)
Comment (Erixonich) [reply modules/core/src/main/java/org/apache/ignite/internal/util/nio/GridTcpNioCommunicationClient.java]: removed

**Review by rpuch — APPROVED** [review]

**Pull request merged**